storage: rework (and remove) refurbishment

tbg opened the pull request:
Instead of trying to 'refurbish' (that is, internally reevaluate and repropose)
Raft commands found to apply at an invalid index, bubble that event up to the
client waiting on the command's 'done' channel.

This is a considerable simplification. A refurbishment was already mostly, but
not quite, a new proposal and is now finally treated as such. This keeps the
evaluation work on the client's goroutine, and various gotchas in the old
implementation simply vanish, along with tne need to treat refurbishments
specially in future work, especially pertaining to proposer-evaluated KV (where
the evaluation step is expensive).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10131)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the use of the term "client". Looks like a nice simplification.

---

Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2347 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbXeQwVjUZWS2a_UOg:-KUbXeQxN_gnly1ql2Ab:bxr1eb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L2347)):*
> ```Go
> 	// Thus we'd risk reproposing a command that has been committed but not yet
> 	// applied.
> 	maxClientShouldRetry := r.mu.state.LeaseAppliedIndex // indexes <= are given up on
> ```

It's slightly confusing that this variable references the client when the retry happens without the client's involvement. Perhaps `s/maxClientShouldRetry/maxShouldRetry/g`.

---

*[pkg/storage/replica.go, line 2364 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbYAjXXlDUS8eC9Iyu:-KUbYAjYD1ADWAdcDWVS:b-qyoem1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L2364)):*
> ```Go
> 		delete(r.mu.proposals, idKey)
> 		// The command's designated lease index range was filled up, so send it
> 		// back to the client for a retry proposal.
> ```

See above comment about the use of "client" being mildly confusing. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-:-KUbYuO6Bv_eZnGvYqs3:b-6wr0zu)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR! Any idea how how to name this type of retry? Agreed that "client retry" is too fuzzy. Reproposal is taken, re-execute is ambiguous, and I'm out of ideas.

---

Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-:-KUbb06p07WXFQrBTdJ0:b-2wcapz)*
<!-- Sent from Reviewable.io -->

petermattis commented:
In some of the comments I think  `s/client retry/retry/g ` is appropriate. `Batch retry`? `Command retry`?

---

Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-:-KUbdnNa3uFNk1p_bBeR:bnit56j)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1486 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbluRZvshIeOltWKrn:-KUbluRZvshIeOltWKro:bqrs9zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1486)):*
> ```Go
> }
> 
> // addWriteCmd is client request's entry point for commands which may mutate
> ```

missing article

---

*[pkg/storage/replica.go, line 1497 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbm2jZYudZgm9ksUSu:-KUbm2j_wQ8Cbl_BFkPQ:b-96ith3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1497)):*
> ```Go
> // - the keys affected by the command are added to the command queue (i.e.
> //   tracked as in-flight mutations).
> // - wait until the command group promises that no overlapping mutations are
> ```

command group? command queue?

---

*[pkg/storage/replica.go, line 1504 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbmAyQDWXbNIW-Y3AD:-KUbmAyQDWXbNIW-Y3AE:bo115az) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1504)):*
> ```Go
> // - a ProposalData is created and inserted into the Replica's in-flight
> //	 proposals map, a lease index is assigned to it, and it is submitted to Raft.
> // - Upon completion, the command is registered in the timestamp cache, removed
> ```

replace "upon completion" with the concrete thing that happens, i.e. something like "wait until raft machinery signals completion of the command via its done channel"

---

*[pkg/storage/replica.go, line 1506 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbmcv0Nc0HLZq4lYv6:-KUbmcv0Nc0HLZq4lYv7:b26w42i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1506)):*
> ```Go
> // - Upon completion, the command is registered in the timestamp cache, removed
> //   from the command queue, and its result (which could be an error) returned
> //   to the client waiting on the channel returned by this method.
> ```

this method doesn't return a channel.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-:-KUbnAJVxpjb27NI4OGV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2705 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUcoXi6tiSe9M5PuNmb:-KUcoXi7gKGOH1X7I_Zt:b-3io3ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c021040823656030718ca0030ba8d3232d568b/pkg/storage/replica.go#L2705)):*
> ```Go
> 			)
> 			cmd.done <- proposalResult{ShouldRetry: true}
> 			close(cmd.done)
> ```

Is it cool to still use this command's context below?  Or should we restore `ctx` to what it was before we set it to `cmd.ctx`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica.go, line 1486 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbluRZvshIeOltWKrn:-KUgw6t8cN03SZKiPyxi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1486)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> missing article
</details>
Done.

---

*[pkg/storage/replica.go, line 1497 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbm2jZYudZgm9ksUSu:-KUgwAs4Xx6lRNAn2kqJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1497)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> command group? command queue?
</details>
Done.

---

*[pkg/storage/replica.go, line 1504 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbmAyQDWXbNIW-Y3AD:-KUgwut-WdnVuPw_aSov:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1504)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> replace "upon completion" with the concrete thing that happens, i.e. something like "wait until raft machinery signals completion of the command via its done channel"
</details>
Done.

---

*[pkg/storage/replica.go, line 1506 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbmcv0Nc0HLZq4lYv6:-KUgwvSUfvnsdoGqfXFa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L1506)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this method doesn't return a channel.
</details>
Done.

---

*[pkg/storage/replica.go, line 2347 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbXeQwVjUZWS2a_UOg:-KUgx1Lz-hpfcavM7qY-:bsrha0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L2347)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> It's slightly confusing that this variable references the client when the retry happens without the client's involvement. Perhaps `s/maxClientShouldRetry/maxShouldRetry/g`.
</details>
`maxMustRetryCommandIndex`

---

*[pkg/storage/replica.go, line 2364 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUbYAjXXlDUS8eC9Iyu:-KUgxBp1LLqTxXh0GIes:b-pcvhzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b90c437308652ca2a510468e898aa075cf664bf3/pkg/storage/replica.go#L2364)):*
>
// The command's designated lease index range was filled up, so send it
// back to the proposer for a retry.

---

*[pkg/storage/replica.go, line 2705 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUcoXi6tiSe9M5PuNmb:-KUh5y7o_HySKEf4WIvP:b-h8o388) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c021040823656030718ca0030ba8d3232d568b/pkg/storage/replica.go#L2705)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Is it cool to still use this command's context below?  Or should we restore `ctx` to what it was before we set it to `cmd.ctx`?
</details>
Nope, not cool. Thanks for catching this one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1518 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUhCEFUY_jIJWtugmt3:-KUhCEFUY_jIJWtugmt4:b-f24y5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb28265a1ac5bdfa00e54054ecfeb5bef816a4ed/pkg/storage/replica.go#L1518)):*
> ```Go
> // - the result of the Raft proposal is read from the channel and the command
> //	 registered with the timestamp cache, removed from the command queue, and
> // 	 its result (which could be an error) returned to the client.
> ```

alignment appears to be off

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: rework (and remove) refurbishment

Instead of trying to 'refurbish' (that is, internally reevaluate and repropose)
Raft commands found to apply at an invalid index, bubble that event up to the
client's goroutine  (waiting on the command's 'done' channel) and retry it
there.

This is a considerable simplification. A refurbishment was already mostly, but
not quite, a new proposal and is now finally treated as such. This keeps the
evaluation work on the client's goroutine, and various gotchas in the old
implementation simply vanish, along with tne need to treat refurbishments
specially in future work, especially pertaining to proposer-evaluated KV (where
the evaluation step is expensive).

tschottdorf commented:
Review status: 6 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/replica.go, line 1518 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUhCEFUY_jIJWtugmt3:-KUhFQ8soWoer5O1SzQ8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb28265a1ac5bdfa00e54054ecfeb5bef816a4ed/pkg/storage/replica.go#L1518)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> alignment appears to be off
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2705 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131#-KUcoXi6tiSe9M5PuNmb:-KUi3cu7LF3QSGvK5aWW:b-1g2hd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1c021040823656030718ca0030ba8d3232d568b/pkg/storage/replica.go#L2705)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Nope, not cool. Thanks for catching this one.
</details>
Smart solution with the defer, nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10131)*
<!-- Sent from Reviewable.io -->
